pkg/controllers/federatedcluster/clusterjoin.go

```diff
@@ -80,232 +75,162 @@ const (
 )
 
 func handleNotJoinedCluster(
+	ctx context.Context,
 	cluster *fedcorev1a1.FederatedCluster,
 	client fedclient.Interface,
 	kubeClient kubeclient.Interface,
 	eventRecorder record.EventRecorder,
 	fedSystemNamespace string,
-	logger klog.Logger,
 	clusterJoinTimeout time.Duration,
-) (*fedcorev1a1.FederatedCluster, error) {
+) (*fedcorev1a1.FederatedCluster, *fedcorev1a1.ClusterCondition, error) {
```

Comment:
why do we return the cluster condition separately from the cluster object?

Reply:
It's to reuse the logic of updating the condition conditionally. If we do it before returning, we'll have to use the function with 11 parameters in a previous commit (see @SOF3 's comment).

Reply:
got it. i just find it rather confusing that while this function mutates certain fields, it also returns a condition that that the caller has to manually set afterwards. could we add a comment to describe this behavior?